### PR TITLE
feat: v0.5.7 — recover terminal width hidden from subprocess + guard upstream rate_limits bug

### LINKED ISSUE #79: Recover terminal width when Claude Code subprocess hides it (no TTY, no COLUMNS)
## Problem

Claude Code spawns the statusLine command as a child process with stdin piped — **there is no TTY and no `COLUMNS` env var**, so:

- `shutil.get_terminal_size()` returns our fallback `(100, 24)` instead of the user's real terminal width
- Our two-stage responsive layout then trims aggressively (compact bucket = 100-149 cols), hiding sections like `rate_limits`, `context_size`, `clock`, `effort`, `version`, `last:N`, `style:N`, `dirs:+N` even when the user has 150-300 cols of room

This means **users with wide terminals see a sparse statusline that's far less informative than it could be**. Reproduction: the user's screenshot at PR #79 shows a 165-col terminal rendering only ~83 chars on Line 2 (5 sections out of 25 available).

## Upstream context

This is tracked in [anthropics/claude-code#22115](https://github.com/anthropics/claude-code/issues/22115) — open since January 2026, heavy community demand (the popular `claude-hud` plugin has 11+ open PRs all addressing the same root cause). **Anthropic has not shipped a fix.** The whole statusline-plugin ecosystem is blocked on this.

We can't wait for Anthropic. We need our own fallback chain.

## Proposed fix

Replace the single `shutil.get_terminal_size()` call with a `_detect_terminal_width()` function that tries each signal in order, returning the first usable value:

1. **`terminal.columns` from stdin JSON** (if Anthropic ever ships this — defensive forward-compat)
2. **`COLUMNS` env var** (some shell wrappers set it; honored by `shutil.get_terminal_size`)
3. **`stty size < /dev/tty`** (works on Linux/macOS/WSL when `/dev/tty` is accessible — most common case)
4. **`tput cols 2>/dev/tty`** (alternate POSIX path)
5. **Existing fallback** `(100, 24)` (last resort, current behavior)

Each step wrapped in try/except so a missing tool / closed `/dev/tty` falls through silently — no crashes, no error output.

## Bundle with: rate_limits epoch-timestamp guard

While we're shipping a width fix, also fix [anthropics/claude-code#52326](https://github.com/anthropics/claude-code/issues/52326) — `rate_limits.five_hour.used_percentage` returns a Unix epoch timestamp (~1.7B) instead of 0/null on fresh sessions. Our current `clamp(0,100)` silently turns this into a false `5h:100% (red)` alarm. Detection: any value > 100 is the upstream bug; treat as "no data yet" → hide the section.

## Acceptance criteria

- A user on a 165-col terminal sees Line 2 fill with rate_limits, speed, context_size, etc. (verified via end-to-end subprocess test)
- All 5 fallback paths have unit tests (mocked, deterministic)
- Each fallback degrades gracefully when its tool is unavailable
- rate_limits section is hidden (not rendered red) when upstream bug returns the timestamp value
- All 332 existing tests still pass
- No new dependencies (stdlib only — `subprocess` is already imported)
- CHANGELOG, README, llms.txt updated

## Related

- Upstream bug: anthropics/claude-code#22115 (terminal width)
- Upstream bug: anthropics/claude-code#52326 (rate_limits epoch)
- Tracked in: #53 (existing umbrella for upstream truncate-related issues)

### PR DESCRIPTION
**Pull request by mkalkere**
Closes #79.

## Summary

Two upstream Anthropic bugs that affect every claude-status user, addressed in one focused PR.

### 1. Recover terminal width that Claude Code hides from the statusLine subprocess

Claude Code spawns the statusLine command with stdin piped — there is **no TTY, no `COLUMNS` env var**, and `shutil.get_terminal_size()` always returns the fallback `(100, 24)`. Result: every user with a terminal wider than 100 cols sees Line 2 silently trimmed even when there's room to spare. The user's screenshot at the issue showed a 165-col terminal rendering only ~83 chars on Line 2, hiding `rate_limits`, `context_size`, `clock`, `effort`, `version`, `last:`, `style:`, `dirs:`.

This is tracked upstream at [anthropics/claude-code#22115](https://github.com/anthropics/claude-code/issues/22115) — open since January 2026, no upstream fix. The whole statusline-plugin ecosystem is blocked on it.

**Fix:** new `_detect_terminal_width()` walks 7 fallback signals until one returns a plausible value:
1. `data["terminal"]["columns"]` from stdin JSON (forward-compat for whenever Anthropic ships it)
2. `COLUMNS` env var
3. `shutil.get_terminal_size()` (handles parent TTY inheritance in some cases)
4. `os.get_terminal_size(fd)` for stderr/stdout/stdin
5. `stty size < /dev/tty`
6. `tput cols 2>/dev/tty`
7. Existing `_COMPACT_LAYOUT_MIN_COLS` fallback

Plausible-range guard `[20, 4000]` rejects garbage from any source. The 4000 upper bound covers ultrawide / 8K / multi-monitor tmux setups that can legitimately exceed 1000 cols.

**Result for the screenshot scenario** (165-col terminal):

| Before | After |
|---|---|
| Line 2: 5 sections, 83 chars | Line 2: **11 sections, 137 chars** |
| missing: rate_limits, context_size, clock, effort, version, last, style, dirs | recovered: 7d:18%, (1000K), last:369h50m, sessions:1, v0.5.7, 21:53 |

### 2. Defensive guard against upstream rate_limits epoch-timestamp bug

[anthropics/claude-code#52326](https://github.com/anthropics/claude-code/issues/52326) — still open. On a fresh 5h or 7d window with no usage data yet, Claude Code returns the `resets_at` epoch timestamp (~1.7e9) in `used_percentage` instead of 0/null. Our previous `clamp(0,100)` silently turned this into a false `5h:100% (red)` alarm on every fresh session for Pro/Max subscribers — telling them they're maxed out when they actually have full headroom.

**Fix:** values >= 1e6 (the actual bug pattern, 7+ orders of magnitude above any plausible percentage) are now treated as "no data yet" and the section is hidden. Values 101-999999 still flow through to the existing clamp — they could be a future Anthropic "overage" indicator above 100% that we shouldn't pre-emptively swallow.

## Diagnostics

`--doctor` now reports both the naive `shutil` width and the detected fallback width side-by-side, so users can verify the recovery worked on their box:

```
Terminal:
  TERM:    xterm-256color
  Columns: 165 (shutil naive: 100)
  Layout:  full (>= 150 cols)
```

## Tests

**352 total** (was 332, +18 new):
- `TestDetectTerminalWidth` (11) — each fallback step pinned independently, boundary values 20/4000, ultrawide 2000, garbage env, non-dict shapes, implausible-input rejection, priority order
- `TestRenderUsesDetectedWidth` (1) — end-to-end at 165 cols proving recovered sections
- `TestRateLimitsEpochTimestampGuard` (6) — 5h epoch hidden, 7d epoch hidden, boundary at 100 (legitimate maxed value passes), 105 (modest over passes — could be future overage), 1e6 (just hits the guard), end-to-end render of bugged value

## Review iterations

4 internal review agents flagged 4 real issues, all addressed before commit:
- **Critical (silent-failure-hunter):** original `> 100` rate-limits guard would silently hide a hypothetical future Anthropic "overage" indicator. Tightened to `>= 1e6` — matches the actual epoch-timestamp bug pattern with zero risk of false positives.
- **Medium (silent-failure-hunter):** `_TERM_WIDTH_MAX = 1000` excluded 8K / multi-monitor tmux setups (some legitimately exceed 1280 cols). Widened to 4000.
- **Doc (comment-analyzer):** docstring said values are "clamped" but the code rejects+falls-through. Corrected.
- **Doc (comment-analyzer):** docstring said `terminal.columns integer` but code accepts numeric strings via `_safe_num`. Corrected.

## Test plan
- [x] All 352 tests pass locally
- [x] Demo + doctor smoke test verified (`Opus 4.7`, layout reporting, dual-width display)
- [x] User's actual scenario (165-col, hydra/main, $879, 18683 lines) verified — Line 2 grows from 5 sections to 11
- [x] No new dependencies (`subprocess` is stdlib)
- [x] No encoding corruption

### CODE REVIEW

File: claude_statusline/cli.py [hunk 1]
```diff
@@ -601,6 +624,145 @@ def append(self, value):
...
+            result = subprocess.run(
+                ["tput", "cols"], stdin=tty,
+                capture_output=True, text=True, timeout=1,
+            )
+        if result.returncode == 0:
+            cols = int(result.stdout.strip())
+            if _TERM_WIDTH_MIN <= cols <= _TERM_WIDTH_MAX:
+                return cols
+    except (OSError, ValueError, subprocess.SubprocessError, FileNotFoundError):
+        pass
```
[the remark this answers is not in the document]
Comment (mkalkere) [reply claude_statusline/cli.py]: Good catch — fixed in 8e5c6cd. Consolidated steps 5+6 into a single `with open("/dev/tty", "r") as tty:` block with both probes nested inside. Bonus benefit: if `/dev/tty` open fails, both probes skip atomically instead of trying twice and burning two except-block trips. Also tightened the inner except to catch `IndexError` explicitly for the parser callable. 352 tests still pass.

File: claude_statusline/cli.py [hunk 2]
```diff
@@ -165,17 +166,39 @@ def _normalize(data):
...
+    # in the legitimate-but-out-of-spec range (e.g. 105) still flow
+    # through to the renderer's existing clamp(0, 100) at the
+    # rate_limits section, so they show as `5h:100% (red)` — which
+    # IS the correct UI for "user is maxed out beyond 100%".
     for period, rl_dict in [("5h", five_h), ("7d", seven_d)]:
-        out["rate_limit_{}_pct".format(period)] = _safe_num(rl_dict.get("used_percentage"))
+        pct = _safe_num(rl_dict.get("used_percentage"))
+        if pct is not None and pct >= 1e6:
+            pct = None
+        out["rate_limit_{}_pct".format(period)] = pct
```
[the remark this answers is not in the document]
Comment (mkalkere) [reply claude_statusline/cli.py:201]: Considered carefully — keeping `>= 1e6`. The two thresholds tension is real:

- **`1e6`**: catches the bug pattern (`~1.7e9` epoch seconds) with **zero risk** of false positives. Anything below 1e6 is preserved for the renderer's existing `clamp(0, 100)` to handle as a normal-but-out-of-spec value.
- **`1e4` (your suggestion)**: tighter, but pre-emptively hides any value 10000-999999. While these aren't currently legitimate semantics, they could be a future Anthropic 'overage' indicator we'd silently swallow.

The silent-failure-hunter agent in our internal review flagged the same trade-off going the other direction: my original `> 100` was too aggressive because it would have silently hidden any future `105` or `200` value. We landed on `1e6` because it's unambiguously the bug pattern (7 orders of magnitude above any plausible percentage) — anything in the `1e4-1e6` band is so vanishingly rare for either bug variant or legitimate Anthropic semantics that erring on 'preserve future flexibility' is the safer call. If a future variant of the bug emits e.g. `5000` we can revisit then; the cost of being wrong in that direction is just a momentary alarm vs. the cost of being wrong in your direction (silently hiding real signals indefinitely).
